kernel/cClosure.ml

```diff
@@ -1413,7 +1413,15 @@ and knht info e t stk =
       { mark = Whnf; term = FProd (n, mk_clos e t, c, e) }, stk
     | LetIn (n,b,t,c) ->
       { mark = Red; term = FLetIn (n, mk_clos e b, mk_clos e t, c, e) }, stk
-    | Evar ev -> { mark = Red; term = FEvar (ev, e) }, stk
+    | Evar ev ->
+      begin match info.i_cache.i_sigma.evar_expand ev with
+      | Some c -> knht info e c stk
+      | None ->
+        if is_irrelevant info.i_cache.i_mode (info.i_cache.i_sigma.evar_relevance ev) then
+          (mk_irrelevant, skip_irrelevant_stack stk)
+        else
+          { mark = Whnf; term = FEvar (ev, e) }, stk
```

Comment:
So this ensures irrelevent evars are not reduced?

Reply:
Yes.